File: tests/member_access_test.rs
//! Tests for member access (obj.property)
//! These tests are written FIRST following TDD red-green-refactor

use aether::interpreter::Evaluator;
use aether::lexer::Scanner;
use aether::parser::Parser;

/// Helper to evaluate expression
fn eval(source: &str) -> Result<String, String> {
    let mut scanner = Scanner::new(source);
    let tokens = scanner.scan_tokens().map_err(|e| e.to_string())?;

    let mut parser = Parser::new(tokens);
    let program = parser.parse().map_err(|e| e.to_string())?;

    let mut evaluator = Evaluator::new();

    // Execute all but last statement
    for stmt in &program.statements[..program.statements.len().saturating_sub(1)] {
        evaluator.exec_stmt(stmt).map_err(|e| e.to_string())?;
    }

    // Evaluate last statement if it's an expression
    if let Some(last) = program.statements.last() {
        if let aether::parser::ast::Stmt::Expr(expr) = last {
            let value = evaluator.eval_expr(expr).map_err(|e| e.to_string())?;
            return Ok(format!("{}", value));
        }
    }

    Ok("null".to_string())
}

#[test]
fn test_array_length_property() {
    let result = eval(r#"
        let arr = [1, 2, 3, 4, 5]
        arr.length
    "#);
    assert_eq!(result.unwrap(), "5");
}

#[test]
fn test_string_length_property() {
    let result = eval(r#"
        let text = "hello"
        text.length
    "#);
    assert_eq!(result.unwrap(), "5");
}

#[test]
fn test_empty_array_length() {
    let result = eval(r#"
        let arr = []
        arr.length
    "#);
    assert_eq!(result.unwrap(), "0");
}

#[test]
fn test_empty_string_length() {
    let result = eval(r#"
        let text = ""
        text.length
    "#);
    assert_eq!(result.unwrap(), "0");
}

#[test]
fn test_direct_literal_member_access() {
    assert_eq!(eval(r#"[1, 2, 3].length"#).unwrap(), "3");
    assert_eq!(eval(r#""test".length"#).unwrap(), "4");
}

#[test]
fn test_member_access_in_expression() {
    let result = eval(r#"
        let arr = [10, 20, 30]
        arr.length + 5
    "#);
    assert_eq!(result.unwrap(), "8");
}

#[test]
fn test_undefined_property_error() {
    let result = eval(r#"
        let arr = [1, 2, 3]
        arr.notexist
    "#);
    assert!(result.is_err());
    let err = result.unwrap_err();
    eprintln!("Error message: {}", err);
    assert!(err.contains("Property") || err.contains("property") || err.contains("member"));
}

#[test]
fn test_member_access_on_non_object() {
    let result = eval(r#"
        let num = 42
        num.length
    "#);
    assert!(result.is_err());
}

File: src/test/java/com/aether/interpreter/MemberAccessTest.java
package com.aether.interpreter;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.aether.exception.AetherRuntimeException;
import com.aether.lexer.Scanner;
import com.aether.parser.Parser;
import com.aether.parser.ast.Stmt;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/** Tests for member access: array.length, string.length, and property error cases. */
class MemberAccessTest {

  private Evaluator evaluator;

  @BeforeEach
  void setUp() {
    evaluator = Evaluator.withoutStdlib();
  }

  private String eval(String source) {
    List<Stmt> stmts = new Parser(new Scanner(source).scanTokens()).parse();
    for (int i = 0; i < stmts.size() - 1; i++) {
      evaluator.execStmt(stmts.get(i));
    }
    if (!stmts.isEmpty() && stmts.getLast() instanceof Stmt.ExprStmt es) {
      return Builtins.display(evaluator.evalExpr(es.expr()));
    }
    if (!stmts.isEmpty()) {
      evaluator.execStmt(stmts.getLast());
    }
    return "null";
  }

  @Test
  void arrayLengthProperty() {
    assertEquals("3", eval("[1, 2, 3].length"));
  }

  @Test
  void stringLengthProperty() {
    assertEquals("5", eval("\"hello\".length"));
  }

  @Test
  void emptyArrayLength() {
    assertEquals("0", eval("[].length"));
  }

  @Test
  void emptyStringLength() {
    assertEquals("0", eval("\"\".length"));
  }

  @Test
  void directLiteralMemberAccess() {
    assertEquals("2", eval("[10, 20].length"));
  }

  @Test
  void memberAccessInExpression() {
    assertEquals("8", eval("let s = \"hello\"\nlet a = [1, 2, 3]\ns.length + a.length"));
  }

  @Test
  void undefinedPropertyError() {
    assertThrows(AetherRuntimeException.class, () -> eval("[1, 2].nonexistent"));
  }

  @Test
  void memberAccessOnNonObject() {
    assertThrows(AetherRuntimeException.class, () -> eval("42.length"));
  }
}
